### ISSUE TITLE
How to prevent stalling job in worker?

### ISSUE THREAD

**Issue by cymen** [bee-queue/bee-queue#75]
I have a worker that needs to do a fair amount of database/API calls. It's doing the same calls for many items in a loop so it could be broken up.

I found that it is taking too long and stalling. I'm going to increase the `stallInterval` but I wondered how can I make this job not stall without increasing the interval?

If I report progress (in the loop), does that prevent stalling?

Or do I need to go look at say `setInterval` or similar work arounds for not blocking the main loop?

I have to admit, I just assumed I could block the main loop in a worker (as I'm running the workers in their own processes).

**Comment by bradvogel** [comment]
How do you know it's stalling? Are you seeing the `stalled` event? Is node consuming 100% CPU?

Just running a bunch of database calls shouldn't stall the event loop in Node. They are async and the event loop continues running while Node is waiting for the database to respond.

**Comment by cymen** [comment]
I found out by the realizing that a job was running on every server restart. Then I checked redis via redis-cli and saw that queue had a stalled job (I haven't looked into why a restart was necessary to retry the stalled job -- I'll go do that now).

I thought the same too on the async calls. I'm making calls to DB but also AWS in a loop. But good point -- not sure why it is stalling.

**Comment by bradvogel** [comment]
When inspecting redis to see if the job was stalled, are you looking in the `stalling` list? That name is actually misleading - it contains a list of jobs that _might_ be stalling, but haven't actually stalled. The only way to know for sure that a job has stalled is to listen for the `stalled` event (btw, I just documented this here: https://github.com/bee-queue/bee-queue/pull/76).

**Comment by cymen** [comment]
@bradvogel Ah! That is good to know. I realized it was stalled mostly due to every time I restarted, the job would be retried.

I think I figured it out -- most of my use of bee-queue is simple but this worker is a bit more complicated. It takes one job and creates many other jobs based on that job. In the loop to create the jobs, I was not calling `.createJob(...)` with `await`. Maybe that isn't it but it looks like one cause. I've modified that and have it reporting progress now. But I can also try reproducing it -- I will do so but I'll close this for now as fairly certain on the right track.
